### PR TITLE
Implemented DeviceConnector and UsbSerial library for Serial Connection

### LINKED ISSUE #241: Missing a Serial connection library for Android
**Actual Behaviour.**  
Missing a Serial connection library for Android.

**Expected Behaviour**  
We need the library to set up the serial connection with Android.

**Would you like to work on the issue?**
Yes.

### LINKED ISSUE #249: Starting a connection with Arduino once get's connected.
**Actual Behaviour.**  
The Connection functionality with the device needs to be added.

**Expected Behaviour**  
Starting a connection with Arduino once get's connected.

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by jddeep**
Finds #241 and Fixes #249 

**Changes**: Implemented UsbSerial library for Serial Connection

**Checklist**: [Please tick following check boxes with `[x]` if the respective task is completed]
- [x] I have used resources from `strings.xml`, `dimens.xml` and `colors.xml` without hard-coding them
- [x] No modifications done at the end of resource files `strings.xml`, `dimens.xml` or `colors.xml`
- [x] I have reformatted code in every file included in this PR [<kbd>CTRL</kbd>+<kbd>ALT</kbd>+<kbd>L</kbd>]
- [x] My code does not contain any extra lines or extra spaces
- [x] I have requested reviews from other members

### CODE REVIEW

**Review by suresh-gandhi — APPROVED** [review]

**Review by CloudyPadmal — CHANGES_REQUESTED** [review]
Please make some code samples. A PR for a library doesn't make sense if it's not in use.

**Comment by CloudyPadmal** [comment]
Also change the PR body replacing **Fixes** as it doesn't fix it.

**Comment by jddeep** [comment]
> 
> 
> Please make some code samples. A PR for a library doesn't make sense if it's not in use.

Yes @CloudyPadmal I had similar thoughts but recently we had https://github.com/fossasia/neurolab-android/pull/186 merged and there I had raised similar concerns as you here. But nevertheless, we can do one thing we can get this merged as the issue is particularly directed to the missing library. I am looking to make a new PR with some implementation done with the USBSerialInterface class by tonight. Hope that is fine. @CloudyPadmal Moreover with this peers can also start working with the former once this library gets implemented.

**Comment by jddeep** [comment]
@CloudyPadmal Please refer here https://github.com/fossasia/neurolab-android/issues/249 I will be working on this just after #247 gets merged.

**Comment by CloudyPadmal** [comment]
@jddeep , when a mentor requests for a change, please do it. Otherwise we'll be wasting our time reviewing your PRs because you won't be making the requested changes. You could have added a class with some functionality in the time you've typed that long argument.

**Comment by jddeep** [comment]
> @jddeep , when a mentor requests for a change, please do it. Otherwise we'll be wasting our time reviewing your PRs because you won't be making the requested changes. You could have added a class with some functionality in the time you've typed that long argument.

Alright @CloudyPadmal I understand.

**Comment by jddeep** [comment]
@CloudyPadmal Done 👍 Have a look now.

File: app/src/main/java/io/neurolab/main/DeviceConnector.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                    connection = null;
+                    device = null;
+                }
+                if (!keep)
+                    break;
+            }
+        }
+    }
+
+    UsbSerialInterface.UsbReadCallback mCallback = new UsbSerialInterface.UsbReadCallback() { //Defining a Callback which triggers whenever data is read.
```
Comment (suresh-gandhi) [inline app/src/main/java/io/neurolab/main/DeviceConnector.java]: Rearrange the code (variables at the top) and remove the Hungarian notation.

File: app/src/main/java/io/neurolab/main/DeviceConnector.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import com.felhr.usbserial.UsbSerialInterface;
+
+import java.io.UnsupportedEncodingException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DeviceConnector {
+
+    UsbManager usbManager;
+    UsbDevice device;
```
Comment (suresh-gandhi) [inline app/src/main/java/io/neurolab/main/DeviceConnector.java]: declare them private and use getters and setters

File: app/src/main/java/io/neurolab/main/DeviceConnector.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                public void onReceive(Context context, Intent intent) {
+                    if (intent.getAction().equals(ACTION_USB_PERMISSION)) {
+                        boolean granted =
+                                intent.getExtras().getBoolean(UsbManager.EXTRA_PERMISSION_GRANTED);
+                        if (granted) {
+                            connection = usbManager.openDevice(device);
+                            serialPort = UsbSerialDevice.createUsbSerialDevice(device, connection);
+                            if (serialPort != null) {
+                                if (serialPort.open()) { //Set Serial Connection Parameters.
+                                    serialPort.setBaudRate(9600);
```
Comment (suresh-gandhi) [inline app/src/main/java/io/neurolab/main/DeviceConnector.java]: a variable can be made for the baud rate. avoid hardcoding the values.

**Review by suresh-gandhi — CHANGES_REQUESTED** [review]

File: app/src/main/java/io/neurolab/main/DeviceConnector.java (hunk 2 above)
Comment (adityastic) [reply app/src/main/java/io/neurolab/main/DeviceConnector.java]: not all of these members need to be exposed

**Review by suresh-gandhi — APPROVED** [review]

File: app/src/main/java/io/neurolab/main/DeviceConnector.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                data = new String(arg0, "UTF-8");
+                data.concat("/n");
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
+            }
+
+            final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() { //Broadcast Receiver to automatically start and stop the Serial connection.
+                @Override
+                public void onReceive(Context context, Intent intent) {
+                    if (intent.getAction().equals(ACTION_USB_PERMISSION)) {
```
Comment (CloudyPadmal) [inline app/src/main/java/io/neurolab/main/DeviceConnector.java]: We can use switch case statement here instead

File: app/src/main/java/io/neurolab/main/DeviceConnector.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+import com.felhr.usbserial.UsbSerialDevice;
+import com.felhr.usbserial.UsbSerialInterface;
+
+import java.io.UnsupportedEncodingException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class DeviceConnector {
+    public final String ACTION_USB_PERMISSION = "com.hariharan.arduinousb.USB_PERMISSION";
```
Comment (CloudyPadmal) [inline app/src/main/java/io/neurolab/main/DeviceConnector.java]: What is this part about; hariharan arduino?

**Review by CloudyPadmal — CHANGES_REQUESTED** [review]

File: app/src/main/java/io/neurolab/main/DeviceConnector.java (hunk 4 above)
Comment (jddeep) [reply app/src/main/java/io/neurolab/main/DeviceConnector.java]: Sure. Done

File: app/src/main/java/io/neurolab/main/DeviceConnector.java (hunk 5 above)
Comment (jddeep) [reply app/src/main/java/io/neurolab/main/DeviceConnector.java]: Ah! I was taking some inspiration from this guy's Arduino-Android code, he has some great stuff going on. Nevertheless, have corrected the perm constant now for us 👍

File: app/src/main/java/io/neurolab/main/DeviceConnector.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    public void checkConnection(Context context) {
+        this.context = context;
+        HashMap<String, UsbDevice> usbDevices = usbManager.getDeviceList();
+        if (!usbDevices.isEmpty()) {
+            boolean keep = true;
+            for (Map.Entry<String, UsbDevice> entry : usbDevices.entrySet()) {
+                device = entry.getValue();
+                int deviceVID = device.getVendorId();
+                if (deviceVID == 0x2341)//Arduino Vendor ID
+                {
```
Comment (suresh-gandhi) [inline app/src/main/java/io/neurolab/main/DeviceConnector.java]: @jddeep follow Java style, not C+ (brackets)

File: app/src/main/java/io/neurolab/main/DeviceConnector.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+    public void checkConnection(Context context) {
+        this.context = context;
+        HashMap<String, UsbDevice> usbDevices = usbManager.getDeviceList();
+        if (!usbDevices.isEmpty()) {
+            boolean keep = true;
+            for (Map.Entry<String, UsbDevice> entry : usbDevices.entrySet()) {
+                device = entry.getValue();
+                int deviceVID = device.getVendorId();
+                if (deviceVID == 0x2341)//Arduino Vendor ID
```
Comment (suresh-gandhi) [inline app/src/main/java/io/neurolab/main/DeviceConnector.java]: Make a variable for this hex value. Will be easy for refactoring in future

**Review by suresh-gandhi — CHANGES_REQUESTED** [review]

File: app/src/main/java/io/neurolab/main/DeviceConnector.java (hunk 7 above)
Comment (suresh-gandhi) [reply app/src/main/java/io/neurolab/main/DeviceConnector.java]: everything other than this seems good.

**Comment by suresh-gandhi** [comment]
@jddeep everything other than this seems fine from my side.

**Review by suresh-gandhi — APPROVED** [review]

**Review by CloudyPadmal — APPROVED** [review]

**Pull request merged**